Mathlib/Tactic/Sat/FromLRAT.lean

```diff
@@ -0,0 +1,609 @@
+
+/-- Asserts that `¬⟦f⟧_v` implies `p`. -/
+structure Fmla.reify (v : Valuation) (f : Fmla) (p : Prop) : Prop where
+  prop : ¬ v.satisfies_fmla f → p
+
+/-- If `f` is unsatisfiable, and every `v` which agrees with `ps` implies `¬⟦f⟧_v → p`, then `p`.
+Equivalently, there exists a valuation `v` which agrees with `ps`,
+and every such valuation yields `¬⟦f⟧_v` because `f` is unsatisfiable. -/
+theorem Fmla.refute (f : Fmla) (hf : f.proof [])
+  (hv : ∀ v, Valuation.implies v (Fmla.reify v f p) ps 0) : p :=
+  (Valuation.mk_implies [] rfl (hv _)).1 (hf _)
+
+/-- Negation turns AND into OR, so `¬⟦f₁ ∧ f₂⟧_v ≡ ¬⟦f₁⟧_v ∨ ¬⟦f₂⟧_v`. -/
+theorem Fmla.reify_or (h₁ : Fmla.reify v f₁ a) (h₂ : Fmla.reify v f₂ b) :
+  Fmla.reify v (f₁.and f₂) (a ∨ b) := by
+  refine ⟨fun H => by_contra fun hn => H ⟨fun c h => by_contra fun hn' => ?_⟩⟩
+  rcases List.mem_append.1 h with h | h
+  · exact hn $ Or.inl $ h₁.1 fun Hc => hn' $ Hc.1 _ h
+  · exact hn $ Or.inr $ h₂.1 fun Hc => hn' $ Hc.1 _ h
+
+/-- Asserts that `¬⟦c⟧_v` implies `p`. -/
+structure Clause.reify (v : Valuation) (c : Clause) (p : Prop) : Prop where
+  prop : ¬ v.satisfies c → p
+
+/-- Reification of a single clause formula. -/
+theorem Fmla.reify_one (h : Clause.reify v c a) : Fmla.reify v (Fmla.one c) a :=
+  ⟨fun H => h.1 fun h => H ⟨fun | _, List.Mem.head .. => h⟩⟩
+
+/-- Asserts that `¬⟦l⟧_v` implies `p`. -/
+structure Literal.reify (v : Valuation) (l : Literal) (p : Prop) : Prop where
+  prop : v.neg l → p
+
+/-- Negation turns OR into AND, so `¬⟦l ∨ c⟧_v ≡ ¬⟦l⟧_v ∧ ¬⟦c⟧_v`. -/
+theorem Clause.reify_and (h₁ : Literal.reify v l a) (h₂ : Clause.reify v c b) :
+  Clause.reify v (Clause.cons l c) (a ∧ b) :=
+  ⟨fun H => ⟨h₁.1 (by_contra fun hn => H hn.elim), h₂.1 fun h => H fun _ => h⟩⟩
+
+/-- The reification of the empty clause is `True`: `¬⟦⊥⟧_v ≡ True`. -/
+theorem Clause.reify_zero : Clause.reify v Clause.nil True := ⟨fun _ => trivial⟩
+
+/-- The reification of a singleton clause `¬⟦l⟧_v ≡ ¬⟦l⟧_v`. -/
+theorem Clause.reify_one (h₁ : Literal.reify v l a) : Clause.reify v (Clause.nil.cons l) a :=
+  ⟨fun H => ((Clause.reify_and h₁ Clause.reify_zero).1 H).1⟩
+
+/-- The reification of a positive literal `¬⟦a⟧_v ≡ ¬a`. -/
+theorem Literal.reify_pos (h : v n ↔ a) : (Literal.pos n).reify v ¬a := ⟨mt h.2⟩
+
+/-- The reification of a negative literal `¬⟦¬a⟧_v ≡ a`. -/
+theorem Literal.reify_neg (h : v n ↔ a) : (Literal.neg n).reify v a := ⟨h.1⟩
+
+end Sat
+
+namespace Mathlib.Tactic.Sat
+
+/-- The representation of a global clause. -/
+structure Clause where
+  /-- The list of literals as read from the input file -/
+  lits : Array Int
+  /-- The clause expression of type `Clause` -/
+  expr : Expr
+  /-- A proof of `⊢ ctx.proof c`.
+  Note that we do not use `have` statements to cache these proofs:
+  this is literally the proof expression itself. As a result, the proof terms
+  rely heavily on dag-like sharing of the expression, and printing these proof terms
```

Comment:
I'm not really sure how best to approach the problem. The problem with `have` statements is that it will be too easy to bust the limit on maximum depth of bound variables that way. Plus bound variables mean that definitions have to be lifted around, leading to de bruijn arithmetic and making it less likely that subterms can be shared.